storage/umass/umass.c

```diff
@@ -0,0 +1,512 @@
+
+
+typedef struct scsi_cdb10 {
+	uint8_t opcode;
+	uint8_t action : 5;
+	uint8_t misc0 : 3;
+	uint32_t lba;
+	uint8_t misc1;
+	uint16_t length;
+	uint8_t control;
+} __attribute__((packed)) scsi_cdb10_t;
+
+
+typedef struct umass_dev {
+	char buffer[8 * UMASS_SECTOR_SIZE];
+	struct umass_dev *prev, *next;
+	usb_devinfo_t instance;
+	char path[32];
+	int pipeCtrl;
+	int pipeIn;
+	int pipeOut;
+	int id;
+	int tag;
+	unsigned port;
+	uint32_t partOffset;
+	uint32_t partSize;
+	handle_t lock;
+	int valid;
+} umass_dev_t;
+
+
+static struct {
+	umass_dev_t *devices;
+	char stack[UMASS_N_MSG_THREADS][1024] __attribute__ ((aligned(8)));
+	unsigned drvport;
+	unsigned msgport;
+	handle_t lock;
+} umass_common;
+
+
+static const usb_device_id_t filters[] = {
+	/* USB Mass Storage class */
+	{ USBDRV_ANY, USBDRV_ANY, USB_CLASS_MASS_STORAGE, USBDRV_ANY, USBDRV_ANY },
+};
+
+
+static int umass_transmit(umass_dev_t *dev, void *cmd, size_t clen, char *data, size_t dlen, int dir)
+{
+	umass_cbw_t cbw = { 0 };
+	umass_csw_t csw = { 0 };
+	int ret = 0, bytes = 0;
+	int dataPipe;
+
+	if (clen > 16)
+		return -1;
+
+	dataPipe = (dir == usb_dir_out) ? dev->pipeOut : dev->pipeIn;
+	cbw.sig = CBW_SIG;
+	cbw.tag = dev->tag++;
+	cbw.dlen = dlen;
+	cbw.flags = (dir == usb_dir_out) ? UMASS_WRITE : UMASS_READ;
+	cbw.lun = 0;
+	cbw.clen = clen;
+	memcpy(cbw.cmd, cmd, clen);
+
+	if ((ret = usb_transferBulk(dev->pipeOut, &cbw, sizeof(cbw), usb_dir_out)) != sizeof(cbw)) {
+		fprintf(stderr, "umass_transmit: usb_transferBulk OUT failed\n");
+		return -EIO;
+	}
+
+	/* Optional data transfer */
+	if (dlen > 0) {
+		if ((ret = usb_transferBulk(dataPipe, data, dlen, dir)) < 0) {
+			fprintf(stderr, "umass_transmit: umass_transmit data transfer failed\n");
+			return ret;
+		}
+		bytes = ret;
+	}
+
+	if ((ret = usb_transferBulk(dev->pipeIn, &csw, sizeof(csw), usb_dir_in)) != sizeof(csw)) {
+		fprintf(stderr, "umass_transmit: usb_transferBulk IN transfer failed\n");
+		return -EIO;
+	}
+
+	/* Transfered finished, check transfer correctness */
+	if (csw.sig != CSW_SIG || csw.tag != cbw.tag || csw.status != 0) {
+		fprintf(stderr, "umass_transmit: transfer incorrect\n");
+		return -EIO;
+	}
+
+	return bytes;
+}
+
+
+static int umass_check(umass_dev_t *dev)
+{
+	scsi_cdb10_t readcmd = {
+		.opcode = 0x28,
+		.length = htons(0x1)
+	};
+	char testcmd[6] = { 0 };
+	mbr_t *mbr;
+
+	if (umass_transmit(dev, testcmd, sizeof(testcmd), NULL, 0, usb_dir_in) < 0) {
+		fprintf(stderr, "umass_transmit failed\n");
+		return -1;
+	}
+
+
+	/* Read MBR */
+	if (umass_transmit(dev, &readcmd, sizeof(readcmd), dev->buffer, UMASS_SECTOR_SIZE, usb_dir_in) < 0) {
+		fprintf(stderr, "umass_transmit 2 failed\n");
+		return -1;
+	}
+
+	mbr = (mbr_t *)dev->buffer;
+	if (mbr->magic != MBR_MAGIC)
+		return -1;
+
+	/* Read only the first partition */
+	dev->partOffset = mbr->pent[0].start;
+	/* For now, our OS can only handle disks smaller than 4 GiB */
+	dev->partSize = min(mbr->pent[0].sectors, 8388607UL);
```

Comment:
It is supposed to be a maximum number of sectors we can handle, because of `off_t` limit. However, I can see now, that I made a mistake here. It should be `(INT_MAX + 1) / SECTOR_SIZE - 1`, so as far as I understand we could handle disks of 2 GiB at maximum.
Once this PR https://github.com/phoenix-rtos/phoenix-rtos-devices/pull/141 gets merged, we should not need this anymore.

Reply:
It has been merged